src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs

```diff
@@ -727,5 +738,24 @@ private string GetEncryptionKeysHashes()
 
             return result;
         }
+
+        private SemaphoreSlim GetFunctionSecretsLock(string functionName)
+        {
+            // We're only serializing access to secrets per-function, not across all functions,
+            // so we need to ensure we're using a single shared lock per-function.
+            if (!_functionSecretsLocks.TryGetValue(functionName, out SemaphoreSlim value))
```

Comment:
Does calling `GetOrAdd()` on the concurrent dictionary not work here?

Reply:
Yes it does, I've made the update. Initially, I was concerned with ensuring that all threads use the same value, and had some questions on how the valueFactory works internally.